utils/pattern/src/frontend/serde.rs

```diff
@@ -5,70 +5,163 @@
+}
+
+impl<B: PatternBackend> Default for PatternString<B>
+where
+    B: PatternBackend,
+    for<'a> B::PlaceholderKeyCow<'a>: core::str::FromStr,
+    for<'a> <B::PlaceholderKeyCow<'a> as core::str::FromStr>::Err: core::fmt::Debug,
+{
+    fn default() -> Self {
+        #[allow(clippy::unwrap_used)] // todo does this work?
+        Self(Pattern::<B>::try_from_str("", Default::default()).unwrap())
+    }
+}
+
+#[cfg(feature = "serde")]
+impl<'de, B: PatternBackend> serde::Deserialize<'de> for PatternString<B>
+where
+    B: PatternBackend,
+    B::PlaceholderKeyCow<'de>: core::str::FromStr,
+    <B::PlaceholderKeyCow<'de> as core::str::FromStr>::Err: core::fmt::Debug,
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        Ok(Self(
+            Pattern::<B>::try_from_str(&<Cow<str>>::deserialize(deserializer)?, Default::default())
+                .map_err(<D::Error as ::serde::de::Error>::custom)?,
+        ))
+    }
+}
+
+#[doc(hidden)]
+pub fn deserialize_borrowed_cow<'de, E, B, D>(
+    deserializer: D,
+) -> Result<Cow<'de, Pattern<B>>, D::Error>
+where
+    D: Deserializer<'de>,
+    B: PatternBackend<StoreFromBytesError = E, Store = str>,
+    B::PlaceholderKeyCow<'de>: Deserialize<'de>,
+{
+    if deserializer.is_human_readable() {
+        Box::<Pattern<B>>::deserialize(deserializer).map(Cow::Owned)
+    } else {
+        <&Pattern<B>>::deserialize(deserializer).map(Cow::Borrowed)
+    }
+}
+
+impl<'de, 'data, B, E> Deserialize<'de> for Box<Pattern<B>>
 where
     'de: 'data,
-    B: PatternBackend<StoreFromBytesError = E>,
-    B::Store: ToOwned + 'de,
-    &'de B::Store: Deserialize<'de>,
+    B: PatternBackend<StoreFromBytesError = E, Store = str>,
     B::PlaceholderKeyCow<'data>: Deserialize<'de>,
-    Store: TryFrom<Cow<'data, B::Store>> + AsRef<B::Store>,
-    Store::Error: Display,
 {
     fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
     where
         D: Deserializer<'de>,
     {
         if deserializer.is_human_readable() {
-            let pattern_items = <HumanReadablePattern<B>>::deserialize(deserializer)?;
-            let pattern_owned: Pattern<B, <B::Store as ToOwned>::Owned> =
-                Pattern::try_from_items(pattern_items.into_iter())
-                    .map_err(<D::Error as ::serde::de::Error>::custom)?;
-            let pattern: Pattern<B, Store> = Pattern::from_store_unchecked(
-                Cow::<B::Store>::Owned(pattern_owned.take_store())
-                    .try_into()
-                    .map_err(<D::Error as ::serde::de::Error>::custom)?,
-            );
-            Ok(pattern)
+            Pattern::<B>::try_from_items(
+                <HumanReadablePattern<B>>::deserialize(deserializer)?.into_iter(),
+            )
+        } else {
+            let store = Box::<B::Store>::deserialize(deserializer)?;
+            B::validate_store(&store).map(|()| Pattern::<B>::from_boxed_store_unchecked(store))
```

Comment:
Suggestion (optional): why not `Pattern::<B>::try_from_boxed_store` (adding that method if necessary)

Reply:
Because that's the only place where this would be needed, it's fairly trivial, and I don't want a `#[cfg(feature = "serde")]` outside this file.